calyx-py/calyx/py_ast.py

```diff
@@ -317,13 +440,22 @@ class Group(Structure):
     connections: list[Connect]
     # XXX: This is a static group now. Remove this and add a new StaticGroup class.
     static_delay: Optional[int] = None
+    attributes: list[GroupAttribute] = field(default_factory=list)
+    loc: Optional[int] = field(default_factory=PosTable.determine_source_loc)
 
     def doc(self) -> str:
-        static_delay_attr = (
-            "" if self.static_delay is None else f'<"promotable"={self.static_delay}>'
+        # hack - add static delay on the fly. Might be problematic if the group was ever going to be written multiple times?
```

Comment:
why is this now a hack? seems to do the exact same thing as before.

Reply:
If `doc()` is called twice (will most likely not happen, but it's an edge case), the static delay attribute would be duplicated in `self.attributes`. I changed all `self.attributes` into sets so this should be resolved now!